app/src/main/java/io/gnosis/safe/ui/transaction/list/TransactionViewListAdapter.kt

```diff
@@ -0,0 +1,47 @@
+package io.gnosis.safe.ui.transaction.list
+
+import android.view.LayoutInflater
+import android.view.ViewGroup
+import androidx.paging.PagingDataAdapter
+import androidx.recyclerview.widget.DiffUtil
+import io.gnosis.safe.ui.base.BaseFactory
+import io.gnosis.safe.ui.base.UnsupportedViewType
+import io.gnosis.safe.ui.transaction.BaseTransactionViewHolder
+import io.gnosis.safe.ui.transaction.TransactionView
+
+class TransactionViewListAdapter(
+    private val factory: BaseFactory<BaseTransactionViewHolder<TransactionView>, TransactionView>
+) : PagingDataAdapter<TransactionView, BaseTransactionViewHolder<TransactionView>>(COMPARATOR)  {
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseTransactionViewHolder<TransactionView> =
+        factory.newViewHolder(
+            factory.layout(LayoutInflater.from(parent.context), parent, viewType),
+            viewType
+        )
+
+    override fun onBindViewHolder(holder: BaseTransactionViewHolder<TransactionView>, position: Int) {
+        val uiModel = getItem(position)
+        holder.bind(uiModel!!, listOf())
+    }
+
+    override fun getItemViewType(position: Int): Int {
+        val uiModel = getItem(position)
+        return uiModel?.let {
+            factory.viewTypeFor(it).takeUnless { it < 0 } ?: throw UnsupportedViewType()
+        } ?: throw UnsupportedViewType()
```

Comment:
Maybe we shouldn't crash here? Not sure though.

Reply:
it should never occur in the first place